manpages/s2ram.8
'\" t
.\"     Title: s2ram
.\"    Author: Tim Dijkstra <tim@famdijkstra.org>
.\"   Updated: Rodolfo García Peñas (kix) <kix@kix.es>
.\"      Date: Feb 29, 2012
.\"    Manual: s2ram
.\"    Source: suspend-utils
.\"  Language: English
.\"
.TH "S2RAM" "8" "Feb 29, 2012" "suspend-utils" "s2ram"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
s2ram \- program to suspend to ram (suspend)
.SH "SYNOPSIS"
.HP \w'\fBs2ram\fR\ 'u \fBs2ram\fR [\-h,\ \-\-help] [\-V,\ \-\-version] [\-n,\ \-\-test] [\-i,\ \-\-identify] [\-\-nofbsuspend] [\-\-force] [\-\-vbe_save] [\-\-vbe_mode] [\-\-radeontool] [\-\-acpi_sleep\ \fInr\fR] [\-\-pci_save]
.SH "DESCRIPTION"
.PP
This manual page documents briefly the \fBs2ram\fR command\&.
.PP
\fBs2ram\fR is a program that will suspend your system to ram (put it in S3 mode)\&. Which means that the system is in a lower power mode\&. It asked to do so it will wake up farely quickly\&. However, if the system will run out of energy entirely you will lose data\&. That is why it is recommended to use \fBs2both\fR(8) \fBs2ram\fR is only provided to find out any work around needed to make suspend to both work\&. See README\&.s2ram\-whitelist and README\&.Debian for more information\&.
.SH "OPTIONS"
.PP
\fB\-V, \-\-version\fR
.RS 4
Version information\&.
.RE
.PP
\fB\-a, \-\-acpi_sleep \fR\fB\fInr\fR\fR
.RS 4
set the acpi_sleep parameter before suspend\&. 1=s3_bios, 2=s3_mode, 3=both\&. Only for Intel platform (x86/amd64)\&.
.RE
.PP
\fB\-f, \-\-force\fR
.RS 4
force suspending, even on unknown machines\&. the following options are only available with \-\-force\&.
.RE
.PP
\fB\-h, \-\-help\fR
.RS 4
Short help text\&.
.RE
.PP
\fB\-i, \-\-identify\fR
.RS 4
prints a string that identifies the machine\&.
.RE
.PP
\fB\-k, \-\-kmstest\fR
.RS 4
test if the kernel supports KMS (Kernel Mode Set)\&. Returns 0 if KMS is supported\&.
.RE
.PP
\fB\-m, \-\-vbe_mode\fR
.RS 4
get VBE mode before suspend and set it after resume\&. Only for Intel platform (x86/amd64)\&.
.RE
.PP
\fB\-n, \-\-test\fR
.RS 4
test if the machine is in the database\&. Returns 0 if known and supported\&.
.RE
.PP
\fB\-p, \-\-vbe_post\fR
.RS 4
VBE POST the graphics card after resume\&. Only for Intel platform (x86/amd64)\&.
.RE
.PP
\fB\-r, \-\-radeontool\fR
.RS 4
turn off the backlight on radeons before suspending\&. Only for Intel platform (x86/amd64)\&.
.RE
.PP
\fB\-s, \-\-vbe_save\fR
.RS 4
save VBE state before suspending and restore after resume\&. Only for Intel platform (x86/amd64)\&.
.RE
.PP
\fB\-v, \-\-pci_save\fR
.RS 4
Save the PCI config space of the VGA card before suspend and restore it after resume\&. Only for Intel platform (x86/amd64)\&.
.RE
.PP
\fB\-\-nofbsuspend\fR
.RS 4
do not suspend the framebuffer (for debugging purposes)\&. Only for Intel platform (x86/amd64)\&.
.RE
.SH "SEE ALSO"
.PP
s2both (8)\&.
.PP
For more information see the README\&.s2ram\-whitelist.
.SH "AUTHOR"
.PP
This manual page was written by Tim Dijkstra <tim@famdijkstra>\&.org for the Debian(TM) system (but may be used by others)\&. Was updated by Rodolfo García <kix@kix.es>. Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 2 any later version published by the Free Software Foundation\&.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2006 Tim Dijkstra
Copyright \(co 2012 Rodolfo García (kix)
.br


s2ram.c
/*
 * Suspend-to-RAM
 *
 * Copyright 2006 Pavel Machek <pavel@suse.cz>
 * Copyright 2011 Rodolfo García Peñas (kix) <kix@kix.es>
 * Distribute under GPLv2.
 */

#include <ctype.h>
#include "config.h"
#include <sys/stat.h>
#include <sys/types.h>
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>

#ifndef S2RAM
#define S2RAM
#endif
#include "vt.h"
#include "s2ram.h"
#include "config_parser.h"

static char s2ram_force;
static char s2ram_quirks[MAX_STR_LEN];

static struct config_par s2ram_parameters[] = {
	{
		.name = "s2ram_force",
		.fmt = "%c",
		.ptr = &s2ram_force,
	},
	{
		.name = "s2ram_quirks",
		.fmt = "%s",
		.ptr = s2ram_quirks,
		.len = MAX_STR_LEN
	},
	{
		.name = NULL,
		.fmt = NULL,
		.ptr = NULL,
		.len = 0,
	}
};


int s2ram_check_kms(void)
{
	DIR *sysfs_dir;
	struct dirent *dentry;
	int ret = -ENOENT;

	sysfs_dir = opendir("/sys/class/drm");
	if (!sysfs_dir)
		return errno;

	while ((dentry = readdir(sysfs_dir))) {
		if (!strncmp(dentry->d_name, "card0-", 6)) {
			ret = 0;
			break;
		}
	}

	closedir(sysfs_dir);

	return ret;
}


/* Actually enter the suspend. May be ran on frozen system. */
int s2ram_generic_do(void)
{
	int ret = 0;
	FILE *f = fopen("/sys/power/state", "w");
	if (!f) {
		printf("/sys/power/state does not exist; what kind of ninja mutant machine is this?\n");
		return ENODEV;
	}
	if (fprintf(f, "mem") < 0) {
		ret = errno;
		perror("s2ram_do");
	}
	/* usually only fclose fails, not fprintf, so it does not matter
	 * that we might overwrite the previous error.
	 */
	if (fclose(f) < 0) {
		ret = errno;
		perror("s2ram_do");
	}
	return ret;
}

/* Parse the configuration file */
int get_s2ram_config(void)
{
	int error;
	char *conf_name = CONFIG_FILE;
	struct stat stat_buf;
	char *ret_name = NULL;

	if (stat(conf_name, &stat_buf)) {
		fprintf(stderr, "%s: Could not stat configuration file\n", conf_name);
		return -ENOENT;
	} else {
		error = parse(ret_name, conf_name, s2ram_parameters);
		if (error) {
			fprintf(stderr, "%s: Could not parse config file\n", ret_name);
			return error;
		}
		if (s2ram_force == 'y') {
			force = 1;

			if (strlen(s2ram_quirks) != 0) {
				printf("Using s2ram_quirks %s\n", s2ram_quirks);
				s2ram_add_quirks(s2ram_quirks);
			}
			return 0;
		}
		return -1;
	}
}

void s2ram_add_quirks(char *s2ram_quirks)
{
	int i;

	/* Move to lowercase */
	for (i=0; s2ram_quirks[i]; i++)
		s2ram_quirks[i] = tolower(s2ram_quirks[i]);

	if (strstr(s2ram_quirks, "vbe_post") != NULL)
		flags |= VBE_POST;
	if (strstr(s2ram_quirks, "vbe_save") != NULL)
		flags |= VBE_POST;
	if (strstr(s2ram_quirks, "vbe_mode") != NULL)
		flags |= VBE_MODE;
	if (strstr(s2ram_quirks, "radeon_off") != NULL)
		flags |= RADEON_OFF;
	if (strstr(s2ram_quirks, "pci_save") != NULL)
		flags |= PCI_SAVE;
	if (strstr(s2ram_quirks, "s3_bios") != NULL)
		flags |= S3_BIOS;
	if (strstr(s2ram_quirks, "s3_mode") != NULL)
		flags |= S3_MODE;
	if (strstr(s2ram_quirks, "fb_nosuspend") != NULL)
		fb_nosuspend = 1;
}
